### PR TITLE
Demand per day billing determinant

### PR DESCRIPTION
**Pull request by jrogers97**
Adds a 'demand per day' billing determinant - where the monthly determinant is the sum of the max demand for every day in that month. New test data is taken from [this sheet](https://docs.google.com/spreadsheets/d/1YbKH6BIhUxi0pLZ2jE74lTX0v9K4jDDLSsXqvtiv_ec/edit#gid=1313587342).

### CODE REVIEW

File: src/rateEngine/LoadProfile.ts [hunk 1]
```diff
@@ -97,6 +99,20 @@ class LoadProfile {
     });
   }
 
+  maxPerDayByMonth(): Array<number> {
```
Comment (jrogers97) [inline src/rateEngine/LoadProfile.ts:102]: I'm open to a better name for this method if one stands out @bknoles @euggra

File: src/rateEngine/LoadProfile.ts [hunk 2]
```diff
@@ -97,6 +99,20 @@ class LoadProfile {
...
+    const months = times(12, (i) => i);
+    const expanded = this.expanded();
+
+    return months.map((m) => {
+      const monthLoads = expanded.filter(({ month }) => m === month);
+
+      // chunk monthly loads by day (31-element array for January, etc.)
+      const dayLoads = Object.values(groupBy(monthLoads, (val) => val.date));
```
Comment (jrogers97) [inline src/rateEngine/LoadProfile.ts:110]: This was an interesting problem. I originally was manually chunking every month into 24-element arrays but forgot that DST adds some complexity. This solution partitions the month by each reading's `date` field, which is definitely safer

File: src/rateEngine/LoadProfile.ts (hunk 1 above)
Comment (euggra) [reply src/rateEngine/LoadProfile.ts:102]: That's a great name.

File: src/rateEngine/LoadProfile.ts (hunk 2 above)
Comment (euggra) [reply src/rateEngine/LoadProfile.ts:110]: Good eye!

File: src/rateEngine/BillingDeterminantFactory.ts [hunk 3]
```diff
@@ -11,25 +11,28 @@ import SurchargeAsPercent, { SurchargeAsPercentArgs } from './billingDeterminant
...
-  'BlockedTiersInMonths' |
-  'FixedPerDay' |
-  'FixedPerMonth' |
-  'MonthlyEnergy' |
-  'MonthlyDemand' |
-  'SurchargeAsPercent' |
-  'HourlyEnergy' |
-  'DemandTimeOfUse' |
-  'DemandTiersInMonths';
+  | 'EnergyTimeOfUse'
```
Comment (bknoles) [inline src/rateEngine/BillingDeterminantFactory.ts:17]: looks like we need to set up linting/prettier here! (def not necessary for this PR)

File: src/rateEngine/BillingDeterminantFactory.ts [hunk 4]
```diff
@@ -60,6 +63,8 @@ class BillingDeterminantFactory {
         return new DemandTiersInMonths(args as BlockedTiersArgs, loadProfile);
       case 'DemandTimeOfUse':
         return new DemandTimeOfUse(args as DemandTimeOfUseArgs, loadProfile);
+      case 'DemandPerDay':
```
Comment (bknoles) [inline src/rateEngine/BillingDeterminantFactory.ts:66]: good job remembering to add this to the factory! without naming names (cough... me...cough), we've definitely forgotten that a few times

File: src/rateEngine/utils/assumptions.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export const DAYS_PER_MONTH = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
```
Comment (bknoles) [inline src/rateEngine/utils/assumptions.ts:1]: as you mentioned above, this is actually not accurate for leap years... should we make this a function that accepts a year? and do we actually have an existing leap year bug in the billing determinants that use this?

File: src/rateEngine/LoadProfile.ts (hunk 1 above)
Comment (bknoles) [reply src/rateEngine/LoadProfile.ts:102]: I see why this gives you pause... it's missing the "sum" concept. something like `maxPerDaySummedByMonth` is more accurate. That's a pretty awkward name, which makes me think there's probably another class we could add.

Something that would yield an API like:

```
loadProfile.max(by: 'day').sum(by: 'month');
// or
loadProfile.maxByDay().sumByMonth();
```

That's out of the scope of this PR, but i would be in favor of using the awkward name until we can take a longer look: `maxPerDaySummedByMonth`.

**Review by bknoles — CHANGES_REQUESTED** [review]
Can you rename the load profile method?

If you think it would be easy to account for DST in our DAYS_PER_MONTH, then that too. If it's not trivial we'll just merge without it.

**Comment by bknoles** [comment]
actually, jake, i wonder if it would be best to put the `maxPerDaySummedByMonth` logic directly in the billing determinant instead of in the load profile.

i love that you added it to the load profile... that's the right instinct, and to me it really calls out the need for a different abstraction that we don't currently have.

so rather than committing to a long public method name we don't like, we could "hide" the logic in the billing determinant and only add it as a public method on the load profile once we have an API that we feel good about

**Comment by jrogers97** [comment]
@bknoles I moved the max demand by day logic into the billing determinant, and adjusted `DAYS_PER_MONTH` to be a function that takes a year (and assumes not a leap year if one isn't provided, since one billing determinant doesn't have a load profile to reference).

I also added a quick test for daysPerMonth, which might be unnecessary but I figured it doesn't hurt to have

**Review by bknoles — APPROVED** [review]
looks great!!

**Pull request merged**